File: 1. Two Sum.py

##def twoSum(nums, target):
##    """
##    :type nums: List[int]
##    :type target: int
##    :rtype: List[int]
##    """
##
##
##    
##    
##    for i in range(len(nums)):
##        for ii in range(i, len(nums)):
##            if i == ii: continue
##            if nums[i] + nums[ii] == target:
##                return [i, ii]
##        
##[2, 6, 7, 9]  t = 13
##
##{ 2 : 0, 6: 1, 7: 2, 9: 3]



def twoSum(nums, target):
    diff = {nums[i] : i for i in range(len(nums))}
    for i in range(len(nums)):
        value = target - nums[i]
        if diff.get(value, False):
            return [i, diff[value]]
    return "dfguck"


print(twoSum([2, 7, 11, 15], 9))
print(twoSum([3, 2, 4], 6))
print(twoSum([3, 3], 6))

File: 1. Two Sum.cc
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> differenceToIndex;

        for (auto i = nums.begin(); i != nums.end(); ++i) {
            differenceToIndex[target - *i] = i - nums.begin();
        }

        for (auto i = nums.begin(); i != nums.end(); ++i) {
            auto found = differenceToIndex[*i];
            int index = i - nums.begin();
            if (found && found != index) {            
                return {found, index};
            }
        }

        return {};

    }
};


int main() {
    vector<int> nums = {1,3,4,2};

    Solution* solution = new Solution();
    vector<int> output = solution->twoSum(nums, 6);

    for (int i = 0; i < 10; ++i) {
        cout << '-';
    }
    cout << endl;

    for (auto x : output) {
        cout << x << endl;
    }

}